bridge/src/message_to_mainnet.rs

```diff
@@ -0,0 +1,109 @@
+use ethereum_types::{Address, U256, H256};
+use contracts::foreign::events::Withdraw;
+use web3::types::Log;
+use ethabi;
+use error::{Error, ErrorKind};
+
+/// the message that is relayed from side to main.
+/// contains all the information required for the relay.
+/// validators sign off on this message.
+#[derive(PartialEq, Debug)]
+pub struct MessageToMainnet {
+	pub recipient: Address,
+	pub value: U256,
+	pub sidenet_transaction_hash: H256,
+	pub mainnet_gas_price: U256,
+}
+
+/// length of a `MessageToMainnet.to_bytes()` in bytes
+pub const MESSAGE_LENGTH: usize = 116;
+
+impl MessageToMainnet {
+	/// parses message from a byte slice
+	pub fn from_bytes(bytes: &[u8]) -> Self {
+		assert_eq!(bytes.len(), MESSAGE_LENGTH);
+
+		Self {
+			recipient: bytes[0..20].into(),
+			value: U256::from_little_endian(&bytes[20..52]),
+			sidenet_transaction_hash: bytes[52..84].into(),
+			mainnet_gas_price: U256::from_little_endian(&bytes[84..MESSAGE_LENGTH]),
+		}
+	}
+
+	/// construct a message from a `Withdraw` event that was logged on `foreign`
+	pub fn from_log(web3_log: Log) -> Result<Self, Error> {
+		let ethabi_raw_log = ethabi::RawLog {
+			topics: web3_log.topics,
+			data: web3_log.data.0,
+		};
+		let withdraw_log = Withdraw::default().parse_log(ethabi_raw_log)?;
+		let hash = web3_log.transaction_hash
+			.ok_or(Error::from_kind(ErrorKind::Msg("`log` must be mined and contain `transaction_hash`".into())))?;
+		Ok(Self {
+			recipient: withdraw_log.recipient.0.into(),
```

Comment:
this conversion is probably redundant since `web3` and `ethabi` are using the same version of `ethereum-types` crate

Reply:
done